### ISSUE TITLE
[REVIEW]: OpenCMP: An Open-Source Computational Multiphysics Package

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3742]
**Submitting author:** @nasserma (<a href="http://orcid.org/0000-0002-1772-0376">Nasser Mohieddin Abukhdeir</a>)
**Repository:** <a href="https://github.com/uw-comphys/opencmp" target ="_blank">https://github.com/uw-comphys/opencmp</a>
**Version:** 1.0
**Editor:** @lucydot
**Reviewer:** @bonh, @WilkAndy
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/efe19348de7efe519e1db433f5692960/status.svg)](https://joss.theoj.org/papers/efe19348de7efe519e1db433f5692960)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/efe19348de7efe519e1db433f5692960"><img src="https://joss.theoj.org/papers/efe19348de7efe519e1db433f5692960/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/efe19348de7efe519e1db433f5692960/status.svg)](https://joss.theoj.org/papers/efe19348de7efe519e1db433f5692960)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/uw-comphys/opencmp" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@bonh & @WilkAndy, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @lucydot know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @bonh
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/uw-comphys/opencmp">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/uw-comphys/opencmp">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@nasserma) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @WilkAndy
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/uw-comphys/opencmp">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/uw-comphys/opencmp">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@nasserma) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.29 s (370.5 files/s, 58814.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          66           2737           5300           6037
reStructuredText                39            860            743           1407
Markdown                         2             24              0             19
make                             1              4              6              9
-------------------------------------------------------------------------------
SUM:                           108           3625           6049           7472
-------------------------------------------------------------------------------


Statistical information for the repository '15e27ad81d067e4382f138f9' was
gathered on 2021/09/21.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Alex Vasile                      1          1992           1031           12.42
elizabethmonte                   6         17212           4099           87.58

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Alex Vasile                1992          100.0          0.0               24.60
elizabethmonte            12082           70.2          1.4               25.00
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3742 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by lucydot** [comment]
@whedon generate pdf from branch publications

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03742/joss.03742/10.21105.joss.03742.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03742/joss.03742/10.21105.joss.03742.pdf) :page_facing_up: :point_left:

**Comment by WilkAndy** [comment]
@nasserma , @elizabethmonte and @Alex-Vasile - i have a question about authorship.  You can read the JOSS guidelines here: https://joss.readthedocs.io/en/latest/submitting.html#authorship .  Given those guidelines, are you OK with James and Nasser being co-authors, even though they appear to have contributed zero code?

**Comment by WilkAndy** [comment]
@nasserma , since i am a heavy contributor to [MOOSE](https://mooseframework.org) i don't like your statement "Open-source packages ... such as MOOSE ... have very steep learning curves and are inaccessible to the general
scientific and engineering communities."   MOOSE was designed to be easy to use.  Perhaps it has changed into something more complicated over time.  Anyway, your statement is just your opinion.  Feel free to make no changes, but I guess i'd tone it down a little, by perhaps changing to "... can have steep learning curves." (without the "inaccessible" part).

**Comment by WilkAndy** [comment]
@nasserma , i'm going on holiday for the next ~10 days, so won't do any more on this for a while, but hopefully you can address the above concerns while i'm away.

Oh, just one final point: your title "multiphysics package" gives the wrong idea to me.  When i think of "multiphysics", i'm thinking of solid mechanics + fluid mechanics (of all kinds) + heat transfer + electromagnetics, etc.  But your code, i *think* is just Navier Stokes.  Can you comment here, or perhaps change the title?

**Comment by whedon** [comment]
:wave: @WilkAndy, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @bonh, please update us on how your review is going (this is an automated reminder).

**Comment by nasserma** [comment]
@lucydot  @WilkAndy This is our first time submitting to JOSS, I just want to confirm how we respond to reviewer comments. Should we be collating them and responding after they are complete or should we respond to each one individually? We have a response ready for @WilkAndy but it also involves revisions to the manuscript.

**Comment by WilkAndy** [comment]
Just trying to install and use.  Please add instructions for getting "NGSolve version 6.2.2101" to your "IO" docs

**Comment by nasserma** [comment]
> @nasserma , since i am a heavy contributor to [MOOSE](https://mooseframework.org) i don't like your statement "Open-source packages ... such as MOOSE ... have very steep learning curves and are inaccessible to the general scientific and engineering communities." MOOSE was designed to be easy to use. Perhaps it has changed into something more complicated over time. Anyway, your statement is just your opinion. Feel free to make no changes, but I guess i'd tone it down a little, by perhaps changing to "... can have steep learning curves." (without the "inaccessible" part).

We have relaxed the language in the revised manuscript, as suggested. Our intention was never to diminish other software packages' utility, but rather differentiate OpenCMP's purpose. Our core mandate is to enable users who are familiar with the command-line interface to perform simulations without having to do any programming whatsoever, while simultaneously empowering developers to contribute new multiphysics models.

**Comment by nasserma** [comment]
> @nasserma , i have not yet read your paper in detail, nor downloaded the software, played with it, etc, but overall i think this paper should be significantly longer. You may have satisfied the minimum requirements for a JOSS paper, but i feel to excite readers to try your software, you need more details. I think you should mention:
> 
>     * code coverage
> 
>     * perhaps outline the tutorials
> 
>     * a bit more details about the user interface
> 
>     * a picture of an example, perhaps explaining why your code/approach is advantageous in terms of usability, accuracy, computational efficiency, etc, compared with other codes.  Sometimes it is hard to not denigrate excellent alternative codes when you're doing this, but you needn't go over-the-top: you're just trying to show the reader your code is nice.
> 
> 
> This is really summarizing work you've already done, so i hope it won't be a huge chore for you.

Based on the JOSS requirements for submissions, "the paper should be between 250-1000 words" and the submitted manuscript is at the upper limit. We could add more details into the paper, but it would be at the expense of other text, most of which has been added to meet the requirements of what the paper should include from the JOSS website,

https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements

We have added references to the OpenCMP website in the revised manuscript. The website content addresses all of your comments except for code coverage, which we have added to the revised manuscript without exceeding the 1000 word limit.

**Comment by Alex-Vasile** [comment]
> @nasserma , @elizabethmonte and @Alex-Vasile - i have a question about authorship. You can read the JOSS guidelines here: https://joss.readthedocs.io/en/latest/submitting.html#authorship . Given those guidelines, are you OK with James and Nasser being co-authors, even though they appear to have contributed zero code?

First, a note about code contributions to the project. The project began on a private internal repo, with commits being periodically being rolled up and copied into the GitHub repo. This results in actual code contributions being unintentionally obscured on the GitHub repo. Future commits will better reflect authorship as development branches are switched from internal repos to GitHub.

To address your question directly, yes I am okay with James and Nasser being co-authors given that they have been active participants. James has contributed code to the repository in the past and is actively implementing new models at the moment. Nasser, as well as James, has provided direction in terms of code structure, and help in understanding, implementing, debugging the numerical methods.

**Comment by bonh** [comment]
Hi @nasserma! 

I am assessing the scholar effort of your code. In my opinion the performance of simulation codes is a very important feature (the most sophisticated implementation of a wonderful physical model will not have a wider impact if the solution time is unacceptable). About the performance of your code (haven't tested it yet): Is it possible to run your code in parallel? In the paper you mention, that utilizing MPI is planned but are you able to use shared memory? Have you compared the performance of your code against other codes? Do you have performed three-dimensional simulations?

**Comment by lucydot** [comment]
@nasserma do you have a timeline for responding to the reviewer comments? I can see that issues have been raised for each one, though no actions yet taken. If you are unsure how best to proceed, please feel free to discuss and ask questions here.

**Comment by WilkAndy** [comment]
Comment: define your physics in Table(1) through equations .  Eg "multi-component reacting NS" might mean lots of things.

Refs https://github.com/uw-comphys/opencmp/issues/9

**Comment by WilkAndy** [comment]
Just trying to install on a fresh windows machine.  Also need:
```
pip install scipy
```

**Comment by WilkAndy** [comment]
I think automated tests are mandatory, and i can't find any.  You have tutorial examples - great, a new user can test things are working to some degree, but it wouldn't take much work to build a test suite out of those, so that new users can be confident things are working as planned.  For instance in "tutorial 4", various norms are outputted, but these are different for my computer than are shown at https://opencmp.io/tutorials/tutorial_4.html .   While i'm not too concerned at the differences, it would be much easier and confidence-inspiring if new users could just write
```
python run_all_tests.py
```
and OpenCMP replies back: "all tests passed".  I use the python `unittest`.

Refs https://github.com/uw-comphys/opencmp/issues/11

**Comment by WilkAndy** [comment]
Various tutorials fail to run.  These could be because i'm running on a vanilla windows system with no extra packages installed.

Refs https://github.com/uw-comphys/opencmp/issues/12
Refs https://github.com/uw-comphys/opencmp/issues/13
Refs https://github.com/uw-comphys/opencmp/issues/14
Refs https://github.com/uw-comphys/opencmp/issues/15

**Comment by WilkAndy** [comment]
I haven't read the documentation carefully, I'm sorry.  However, at the moment i believe all your tutorials are "simple" cases that can be solved easily using other software.  If true, this doesn't really compel me to try your new software.  I believe the big new thing about your software is diffuse boundary method.  How about highlighting that through an awesome example, both in your online documentation and this JOSS paper.

**Comment by nasserma** [comment]
> I think automated tests are mandatory, and i can't find any. You have tutorial examples - great, a new user can test things are working to some degree, but it wouldn't take much work to build a test suite out of those, so that new users can be confident things are working as planned. For instance in "tutorial 4", various norms are outputted, but these are different for my computer than are shown at https://opencmp.io/tutorials/tutorial_4.html . While i'm not too concerned at the differences, it would be much easier and confidence-inspiring if new users could just write
> 
> ```
> python run_all_tests.py
> ```
> 
> and OpenCMP replies back: "all tests passed". I use the python `unittest`.
> 
> Refs [uw-comphys/opencmp#11](https://github.com/uw-comphys/opencmp/issues/11)

Please see the README.md file which describes our unit test procedure. We use pytest instead of unittest for unit and integration tests. We prefer it for a few reasons, but if you see the pytests/ directory there is also a README.md file with further information regarding how (i) how run the existing tests and (ii) how to create new tests.

At this point we have quite a decent amount of tests, they are all present in subdirectories of pytests/ and are commented/documented.

That being said, the most recent push to the Github branch seems to have caused some significant issues with pytests and the tutorials. We are in the process of working out how to deal with our internal VCS and external VCS (Github). I will work this out with the others ASAP.

**Comment by nasserma** [comment]
> Various tutorials fail to run. These could be because i'm running on a vanilla windows system with no extra packages installed.
> 
> Refs [uw-comphys/opencmp#12](https://github.com/uw-comphys/opencmp/issues/12) Refs [uw-comphys/opencmp#13](https://github.com/uw-comphys/opencmp/issues/13) Refs [uw-comphys/opencmp#14](https://github.com/uw-comphys/opencmp/issues/14) Refs [uw-comphys/opencmp#15](https://github.com/uw-comphys/opencmp/issues/15)

Please use revision ID  a475159 for now, there was definitely an issue pushing changes from the last revision and we will resolve in shortly.

**Comment by bonh** [comment]
I will continue my review after we have discussed uw-comphys/opencmp#6 and uw-comphys/opencmp#7. I rate these as critical to assess the scholarly effort.

**Comment by bonh** [comment]
I am sorry for not being more clear: recommending the paper for publication in JOSS and improving the package are two different parts of my work here, I think. So, as I wrote just now in uw-comphys/opencmp#6, 

> in my opinion, a CFD package which is not able to at least simulate moderate scale problems AND does not implement some new methodology does not meet the "scholarly effort" described by JOSS.

I just want to engage in a discussion with you about your work! Now if you are able to simulate moderate size problems (e.g., something in 3D or so, using shared memory) this would be okay for me to proceeded with the review. Same way if you convince me of the novelty of your method (diffuse interface). 
However, I think that if you offer the user MPI (or some other distributed memory parallelization), this would greatly improve your software and therefore the paper!

**Comment by nasserma** [comment]
@bonh @WilkAndy @lucydot Please see the recent revision that we pushed, it addresses many of the comments that were made, although there are a few outstanding issues that I would like to discuss before we address them.

Please see the open issues,

https://github.com/uw-comphys/opencmp/issues

and our responses. Once we have your feedback we'll make the next round of revisions and please note that the installation instructions have changed,

https://opencmp.io/getting_started/installation_guide.html

**Comment by bonh** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #3742 with the following error: 
```
 [WARNING] Could not convert image '/tmp/tex2pdf.-9d76ffa7cd6dcdaa/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg': check that rsvg-convert is in path.
  rsvg-convert: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.-9d76ffa7cd6dcd
aa/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.627 ...b6359d3c5367d353cc654cde8d34a5a41722.svg}

Looks like we failed to compile the PDF

```

**Comment by bonh** [comment]
@whedon generate pdf from branch publications

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3742 with the following error: 
```
 [WARNING] Could not convert image '/tmp/tex2pdf.-949c2ce0db16a535/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg': check that rsvg-convert is in path.
  rsvg-convert: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.-949c2ce0db16a5
35/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.627 ...b6359d3c5367d353cc654cde8d34a5a41722.svg}

Looks like we failed to compile the PDF

```

**Comment by bonh** [comment]
@lucydot @nasserma , am i doing something wrong? Please help me to generate the pdf.

**Comment by nasserma** [comment]
@bonh I will fix this shortly, sorry for the oversight.

**Comment by nasserma** [comment]
@whedon generate pdf from branch publications

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3742 with the following error: 
```
 [WARNING] Could not convert image '/tmp/tex2pdf.-d590ef25176202e3/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg': check that rsvg-convert is in path.
  rsvg-convert: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.-d590ef25176202
e3/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.627 ...b6359d3c5367d353cc654cde8d34a5a41722.svg}

Looks like we failed to compile the PDF

```

**Comment by nasserma** [comment]
@whedon generate pdf from branch publications

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3742 with the following error: 
```
 [WARNING] Could not convert image '/tmp/tex2pdf.-09431f8401f36e88/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg': check that rsvg-convert is in path.
  rsvg-convert: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.-09431f8401f36e
88/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.627 ...b6359d3c5367d353cc654cde8d34a5a41722.svg}

Looks like we failed to compile the PDF

```

**Comment by nasserma** [comment]
I do not understand the issue, it successfully compiles using the preview link,

https://whedon.theoj.org/

using,

https://github.com/uw-comphys/opencmp/

with the branch specified as _publications_

**Comment by lucydot** [comment]
Hi @nasserma yes - you are right - it does compile correctly using the preview link you have provided above.

@bonh @WilkAndy I suggest that as a workaround for compiling the paper you use the link [https://whedon.theoj.org/](https://whedon.theoj.org/]) (with `https://github.com/uw-comphys/opencmp/` as the repo and the branch specified as `publications`).

I'll look into why whedon command `@whedon generate pdf from branch publications` isn't working.

**Comment by lucydot** [comment]
@nasserma @bonh @WilkAndy please note that I am setting my out of office for the next two weeks over Christmas. I will check in once a week, but my responses might be slower than usual for the next fortnight.

**Comment by lucydot** [comment]
@openjournals/devs - pinging you here as we are having problems compiling the paper with the command `@whedon generate pdf from branch publications`. Compiling using the link [https://whedon.theoj.org/](https://whedon.theoj.org/) with `https://github.com/uw-comphys/opencmp/` as the repo and `publications` as the branch is working.

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3742 with the following error: 
```
 [WARNING] Could not convert image '/tmp/tex2pdf.-fc42df7ef731c975/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg': check that rsvg-convert is in path.
  rsvg-convert: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.-fc42df7ef731c9
75/6bfdb6359d3c5367d353cc654cde8d34a5a41722.svg (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.627 ...b6359d3c5367d353cc654cde8d34a5a41722.svg}

Looks like we failed to compile the PDF

```

**Comment by bonh** [comment]
On documentation: uw-comphys/opencmp#18

**Comment by lucydot** [comment]
Hello @nasserma @Alex-Vasile and our reviewers @bonh @WilkAndy -

I'm aware this review has been going for > 3months so thought I'd do a quick summary of where we are at now, and try to establish a timeframe for the review.

Looking at the issues that are open:

- [uw-comphys/opencmp#6](https://github.com/uw-comphys/opencmp/issues/6) relates to distributed memory parallelisation (MPI) and there is a question as to whether this is needed to demonstrate substantial scholarly effort, as stated in the JOSS guidelines [here](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort). @bonh @WilkAndy Note that substantial scholarly effort is addressed in the closed issue [uw-comphys/opencmp#7](https://github.com/uw-comphys/opencmp/issues/7), so this criteria may be met through other means.
- [uw-comphys/opencmp#8](https://github.com/uw-comphys/opencmp/issues/8) relates to the paper. It looks like we are waiting on an example demonstrating substantial scholarly effort to be included. @nasserma are you waiting on further feedback from the reviewers on this item?
- [uw-comphys/opencmp#18](https://github.com/uw-comphys/opencmp/issues/18). This asks for a small adjustment to the documentation.
- [uw-comphys/opencmp#19](https://github.com/uw-comphys/opencmp/issues/19). There is a failing test which is currently under discussion.

@bonh @WilkAndy, with the above in mind, could I ask you to update your tick-boxes (if appropriate) so I can get a feel for where we are at in the review process?

**Comment by bonh** [comment]
Hi,
- uw-comphys/opencmp#6: After I am able to get the software to run on my computer I will check out the 3D simulation case. If it works I consent to close this issue. I am already convinced of the substantial scholarly effort (see the tick-box). 
- uw-comphys/opencmp#18: this is ongoing while I walk through the documentation. Right now I am stuck at testing right after installation, see uw-comphys/opencmp#19.

**Comment by lucydot** [comment]
Hi @bonh @WilkAndy, have either of you been able to succesfully run the tests after installation? As highlighted above, this is the current sticking point. The relevant issue thread (uw-comphys/opencmp#19) has been updated and apt-get has been suggested rather than using conda for install.

**Comment by nasserma** [comment]
Hi @lucydot , @bonh was able to run all of our tests, please see the thread here,

https://github.com/uw-comphys/opencmp/issues/19

there was an issue with the pytest module's parsing of the --runslow option, which is corrected in our documentation. From my understanding we have addressed all of the reviewers' comments and, with our most recent version, now have two tutorials which demonstrate 3D simulations both using the diffuse-interface method and flow past an immersed cylinder.

The last open issue is,

https://github.com/uw-comphys/opencmp/issues/8

but we have addressed all of the issues, apart from adding simulation results images to the paper. We can do this, but since we provide links to the examples/tutorials and images/movies are available through our website, we think that is reasonable unless there is a strong objection.

**Comment by lucydot** [comment]
Hello @nasserma  - thank you for the update. @bonh @WilkAndy if you are able to give us updates on your review in the next fortnight or so that would very useful. I'm aware this review has been taking some time, hopefully now we are at a stage where the functionality and central claims of the software can be assessed. If you have any questions or are unavailable to review over the immediate time period, please let me know.

**Comment by WilkAndy** [comment]
I'm kind of pulling my hair out over here in Australia!!    I'm pretty sure i installed and tested the software, but have since got a new computer and am having lots of trouble installing ngsolve.  It's a real beast with the conda version not working, the mac version not working properly with opencmp, and properly testing opencmp being problematic (for me) on windows.

I'm disappointed that the authors don't want to enhance their paper in various ways i suggested above, because i feel it'd encourage readers to try opencmp, but that's no reason for rejecting the manuscript.   So, i think i "just" have to install ngsolve.

**Comment by bonh** [comment]
I get the feeling that you, @nasserma, and your coauthors find me rather annoying to be honest. If so, I am really sorry, but I think it is my duty to be thorough and a little pushy. Sadly, you seldom engage in a discussion with me.

I would like to reopen uw-comphys/opencmp/issues/18 as I am not done with the documentation yet. Same for uw-comphys/opencmp/issues/6 as I was stuck with testing and did not look at the 3D simulations yet. If you close the issues I loose track, sorry about that.

I am continue now with trying out a few simulations but I am convinced by now that this will be publishable.

@WilkAndy I ended up putting it into a docker container (Ubuntu and then a lot of apt get and pip from inside the container). If you want I can try to share it with you (it is not nice...).

**Comment by nasserma** [comment]
> I'm kind of pulling my hair out over here in Australia!! I'm pretty sure i installed and tested the software, but have since got a new computer and am having lots of trouble installing ngsolve. It's a real beast with the conda version not working, the mac version not working properly with opencmp, and properly testing opencmp being problematic (for me) on windows.
> 
> I'm disappointed that the authors don't want to enhance their paper in various ways i suggested above, because i feel it'd encourage readers to try opencmp, but that's no reason for rejecting the manuscript. So, i think i "just" have to install ngsolve.

Sorry that you are having trouble with the installation using Anaconda and MacOS, we will definitely work with the NGSolve developers to try and find out what issues exist and how to fix them. Similarly to you having issues testing on these platforms, we have issues testing on other platforms since we predominantly use some variation of Linux for desktop and high-performance computing.

Regarding your second comment, if you reread my original response from a few days ago I indicated that we would make this change if there were strong objections. This is something we can complete over the weekend and will push a new version of the manuscript for you to review on Monday.

**Comment by bonh** [comment]
I started to carefully read the paper. Comments are in https://github.com/uw-comphys/opencmp/issues/22.

**Comment by nasserma** [comment]
> I get the feeling that you, @nasserma, and your coauthors find me rather annoying to be honest. If so, I am really sorry, but I think it is my duty to be thorough and a little pushy. Sadly, you seldom engage in a discussion with me.
> 
> I would like to reopen [uw-comphys/opencmp/issues/18](https://github.com/uw-comphys/opencmp/issues/18) as I am not done with the documentation yet. Same for [uw-comphys/opencmp/issues/6](https://github.com/uw-comphys/opencmp/issues/6) as I was stuck with testing and did not look at the 3D simulations yet. If you close the issues I loose track, sorry about that.
> 
> I am continue now with trying out a few simulations but I am convinced by now that this will be publishable.

I am sorry if you get the feeling that I am annoyed with you or am not engaging in conversation. You have to understand, this is a very different peer review process from what I am accustomed to. I am doing my best to adapt to it, but I also must consult with my co-authors before responding, making changes to the code, and revising the manuscript. This takes time, as I am sure you understand.

Regarding the two issues with the code that you found, we are all very sorry for not catching those and will correct this in the future by implement automated testing of the examples, in addition to the unit tests/pytests. These were introduced by some refactoring and it was my responsibility to test the examples. We will add this to our to-do list for the weekend!

Please check back on Monday, and make sure to use the _publications_ branch, which we'll merge into the main branch once all of these issues are resolved.

**Comment by nasserma** [comment]
@lucydot @WilkAndy @bonh I have pushed a revised manuscript with changes meant to address all of the issues and comments from the open issues (#8, 22, 23). I am having trouble with the layout of images, but I want to get an otherwise complete revision out while I sort out using the JOSS docker image to compile locally.

Please let me know what you all think!

**Comment by nasserma** [comment]
@whedon generate pdf from branch publications

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03742/joss.03742/10.21105.joss.03742.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03742/joss.03742/10.21105.joss.03742.pdf) :page_facing_up: :point_left:

**Comment by bonh** [comment]
I read through the paper again and added some new discussion points in https://github.com/uw-comphys/opencmp/issues/22. And there are some old points which have not been addressed yet.

**Comment by nasserma** [comment]
I reopened them.

> > I would like to reopen [uw-comphys/opencmp#18](https://github.com/uw-comphys/opencmp/issues/18) as I am not done with the documentation yet. Same for [uw-comphys/opencmp#6](https://github.com/uw-comphys/opencmp/issues/6) as I was stuck with testing and did not look at the 3D simulations yet. If you close the issues I loose track, sorry about that.
> 
> Perhaps you could reopen them?

**Comment by nasserma** [comment]
I responded to all of your new comments, however, you indicated that a previous comment was not responded to but I could not determine which one you were referring to.

> I read through the paper again and added some new discussion points in [uw-comphys/opencmp#22](https://github.com/uw-comphys/opencmp/issues/22). And there are some old points which have not been addressed yet.

**Comment by nasserma** [comment]
@whedon generate pdf from branch publications

**Comment by nasserma** [comment]
@editorialbot generate pdf from branch publications

**Comment by danielskatz** [comment]
@xuanxu - Can you please look at this error
> /home/runner/work/_actions/xuanxu/paper-action/main/vendor/bundle/ruby/3.1.0/gems/theoj-1.2.1/lib/theoj/paper.rb:155:in `failure': Couldn't check the bibtex because branch name is incorrect: publications (Theoj::Error)

**Comment by xuanxu** [comment]
Error found (ambiguity between branch name and folder name) and hopefully fixed, let's try again.

**Comment by xuanxu** [comment]
@editorialbot generate pdf

**Comment by xuanxu** [comment]
@editorialbot check references

**Comment by nasserma** [comment]
@lucydot @WilkAndy @bonh I have confirmed that the latest PDF is the correct/updated version of the manuscript. Sorry for not catching this earlier, but the preview PDF built correctly and I did not check the Github version thinking that it would be exactly the same.

Since the revised manuscript is significantly different than before, I am hoping that both reviewers can revisit their comments and identify which ones are addressed satisfactorily and which ones are not. Thanks.

**Comment by lucydot** [comment]
@editorialbot check references

**Comment by lucydot** [comment]
Hi @nasserma please could you confirm the missing DOIs listed above are correct, and if so, add them to your paper?

**Comment by nasserma** [comment]
@editorialbot check references

**Comment by bonh** [comment]
@editorialbot generate pdf

**Comment by lucydot** [comment]
Excellent @bonh thank-you very much for reviewing for JOSS.

@WilkAndy please could you give an update on your review?

**Comment by WilkAndy** [comment]
@lucydot and authors - please accept my apologies.  My notifications have been going to spam for some time, and i had forgotten.  Please give me 24 hours to complete my reivew.

**Comment by WilkAndy** [comment]
Suggestion (not mandatory for publication): improve the lighting in your paraview figures - they're super dark at the moment!

**Comment by WilkAndy** [comment]
Well done,  @nasserma @elizabethmonte @Alex-Vasile , on a neat paper and nice software!    (I just have one minor comment (see above) you might like to address before publication - totally up to you.)

**Comment by nasserma** [comment]
@WilkAndy This is good news! We will improve the figures over the next few days and get back to you!

**Comment by lucydot** [comment]
Excellent - thank you for your update @WilkAndy. @nasserma I will wait here for an update comment from yourself or your team before proceeding to the next stage.

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by lucydot** [comment]
@nasserma it looks like there may be an issue with an inline reference. The error message reads `attempt to get length of a nil value (local 'author')`. 

Looking at the bib file, there are "author" fields missing in the "misc" entries, but I don't think these should be causing the problem. Perhaps you need double curly brackets around "Message Passing Interface Forum" as this is not in the standard author format .

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@WilkAndy @lucydot I have improved the figures for tutorials 8 and 10 in the revised manuscript (above), the only issue left are that the figures for tutorial 9 need to be side-by-side. There seem to be a few ways to do this using Markdown, if there is a preferred one please let me know.

**Comment by lucydot** [comment]
Hi @nasserma -  I can see the double brackets around "Message Passing Interface Forum" seemed to fix the problem  - great.

There is not a preferred method for aligning images in markdown that I am aware of. I think the best approach is to implement, then test that it compiles into pdf as expected here in the thread, using the editorialbot as you have been doing. It may be easier to have a single image file with two sub-images rather than aligning in markdown. The images do look rather large at the moment, I suggest that they can all be scaled down for easier readability of the paper as a whole.

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
Hi @lucydot - I have made all of the editorial comment changes and merged the two images together that were scaled poorly in the previous version of the manuscript. Please see the most recent version of the manuscript above, thanks.

**Comment by lucydot** [comment]
@nasserma the latest version of the pdf has two of the images split over two pages. I'm not aware of a programmatic way to stop breaking images over two pages (I didn't know that markdown would allow this in fact) - perhaps you could try moving the position of the images within the main text so that this doesn't happen?

**Comment by lucydot** [comment]
@nasserma a few points that remain from my previous comments:
- 'Nguyen et al. (2018)' should read '(Nguyen et al., 2018)' for consistency
- OpenFOAM and SU2 are linked to directly rather than referenced: can you confirm that references for each package are not available? If references are available I suggest you provide this in the text as done for the other packages.
- the arrow formatting has not worked: 'A&rarrB' is displayed.

Please check after making changes to the markdown that they appear as expected when viewing the compiled pdf. Thanks - Lucy.

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by nasserma** [comment]
@lucydot Please see the latest version which has the following changes,

- Two sets of images were merged so that they will not be split across pages (tutorials 8 and 10 images).
- Updated citation of '(Nguyen et al., 2018)', I was trying to cite to papers in a single citation.
- OpenFOAM and SU2 are cited per instructions from their websites
- arrow formatting corrected

Hopefully this resolves all of the issues!

**Comment by lucydot** [comment]
@nasserma I'm just checking with other editorial team members about citatiion formatting when there is not a date associated with a publication (currently displaying as "n.d").

**Comment by lucydot** [comment]
@nasserma
- the formatting for the MPI Forum author field is not correct - in your bib file the title should be in double curly brackets: "{{Message Parsing Interface Forum}}"
- there seems to be an inconsistency with the "misc" entries in the bib file. A `year` field has been supplied for ngsolve, but not for the other websites. I suggest removing the `year` field for ngsolve (which I think is a little vague in this context - year accessed or published?) and adding to all "misc" fields a `urldate` field in the following format `urldate={2014-04-01}`.

**Comment by nasserma** [comment]
Hi @lucydot , I made the changes specified, but the adding the urldate field to the bib entries does not change the presence of "N.D." in the resulting citations. Also, I have populated that field with the date that the page was updated (extracted from the HTML metadata), but in the article proof it indicates the field should reflect the date accessed. I am pushing an update to the OpenFOAM bib entry now, since that was the only one that was not dynamically created.

**Comment by lucydot** [comment]
Ah @nasserma thanks for trying the `urldate` field. Yes, the `urldate` is the date the page was accessed - if you could update accordingly (to the date accessed) I think we are then ready to recommend to editor in chiefs for acceptance. We'll leave the `n.d.` as it is - I can see there are other JOSS papers that have been published with this notation instead of a date.

**Comment by nasserma** [comment]
@editorialbot generate pdf

**Comment by lucydot** [comment]
@editorialbot commands

**Comment by lucydot** [comment]
@editorialbot check references

**Comment by lucydot** [comment]
Hi @nasserma ok it looks good to me now. 

Please can you do the following:

- Make a tagged release of your software
- Archive this release on Zenodo (or similar)
        - The archive should have exactly the same title and author list as your JOSS paper
- Report the archive DOI and version number in this thread

**Comment by nasserma** [comment]
Hi @lucydot the archive DOI is 10.5281/zenodo.6515912 and version number is v1.0.0

Thanks!

**Comment by lucydot** [comment]
@editorialbot set 10.5281/zenodo.6515912 as archive

**Comment by lucydot** [comment]
@editorialbot set v1.0.0 as version

**Comment by lucydot** [comment]
@editorialbot recommend-accept

**Comment by nasserma** [comment]
Hi @kyleniemeyer, thanks, I merged the corrections.

**Comment by kyleniemeyer** [comment]
@editorialbot generate pdf

**Comment by kyleniemeyer** [comment]
@editorialbot accept

**Comment by kyleniemeyer** [comment]
Congratulations @nasserma on your article's publication in JOSS!

Many thanks to @bonh and @WilkAndy for reviewing this, and @lucydot for editing.

**Comment by nasserma** [comment]
> Congratulations @nasserma on your article's publication in JOSS!
> 
> Many thanks to @bonh and @WilkAndy for reviewing this, and @lucydot for editing.

Thanks to all!

**Comment by tarleb** [comment]
Hi all! We are working towards a more unified archive of all JOSS papers. In this process, I noted that the Markdown sources for this paper contain markup to include an image that does not exist, namely `images/tutorial_8_geom.png`, and is hence not included in the PDF. Judging from the surrounding text, the image may have been combined with another graphic, so this may not be a problem. Otherwise we may want to re-publish this paper with the correct image. @nasserma, could you take a look?

**Comment by nasserma** [comment]
Hi @tarleb , the situation is as you expected and line 106 of the current version of the Markdown source for the paper should be removed. I will push a revision of the file to the OpenCMP repository, but since this does not result in a change in the compiled PDF does anything else need to be done?

**Comment by tarleb** [comment]
Thanks for the feedback. I'm generating JATS from a copy of the paper sources, so I can fix locally. With the PDF looking as intended, there is nothing else we'd need to do here. The rest is on my side. :+1:
